File: app/src/main/java/org/example/example/game/Board.java
package org.example.example.game;


import org.example.example.piece.Color;
import org.example.example.piece.Piece;

import java.util.ArrayList;
import java.util.List;

public class Board {
    private List<Cordinate> cordinates;

    public Board(List<Cordinate> cordinates) {
        this.cordinates = cordinates;
    }

    public List<Cordinate> getCordinates() {
        return cordinates;
    }

    public Cordinate getCordinate(int x, int y) {
        for (Cordinate cordinate : cordinates) {
            if (cordinate.getX() == x && cordinate.getY() == y) {
                return cordinate;
            }
        }
        return null;
    }

    public Cordinate getPieceCordinate(Piece piece) {
        for (Cordinate cordinate : cordinates) {
            if (cordinate.getPiece().equals(piece)) {
                return cordinate;
            }
        }
        return null;
    }

    public Cordinate getPieceCordinate(Piece piece, Color color) {
        for (Cordinate cordinate : cordinates) {
            if (cordinate.getPiece().equals(piece) && cordinate.getPiece().getColor().equals(color)) {
                return cordinate;
            }
        }
        return null;
    }

    public List<Cordinate> getAllCordinatesWithPieces() {
        List<Cordinate> cordinates = new ArrayList<>();
        for (Cordinate cordinate : this.cordinates) {
            if (cordinate.hasPiece()) {
                cordinates.add(cordinate);
            }
        }
        return cordinates;
    }

    public List<Cordinate> getEnemyPiecesCordinates(Color color) {
        List<Cordinate> cordinates = new ArrayList<>();
        for (Cordinate cordinate : this.cordinates) {
            if (cordinate.hasPiece() && cordinate.getPiece().getColor() != color) {
                cordinates.add(cordinate);
            }
        }
        return cordinates;
    }

    public List<Cordinate> getFriendlyPiecesCordinates(Color color) {
        List<Cordinate> cordinates = new ArrayList<>();
        for (Cordinate cordinate : this.cordinates) {
            if (cordinate.hasPiece() && cordinate.getPiece().getColor() == color) {
                cordinates.add(cordinate);
            }
        }
        return cordinates;
    }

    public List<Cordinate> getCordinatesBetween(int x1, int y1, int x2, int y2) {
        List<Cordinate> cordinates = new ArrayList<>();
        if (x1 == x2) {
            for (int i = Math.min(y1, y2) + 1; i < Math.max(y1, y2); i++) {
                cordinates.add(getCordinate(x1, i));
            }
        } else if (y1 == y2) {
            for (int i = Math.min(x1, x2) + 1; i < Math.max(x1, x2); i++) {
                cordinates.add(getCordinate(i, y1));
            }
        } else if (Math.abs(x1 - x2) == Math.abs(y1 - y2)) {
            int x = Math.min(x1, x2) + 1;
            int y = Math.min(y1, y2) + 1;
            while (x < Math.max(x1, x2)) {
                cordinates.add(getCordinate(x, y));
                x++;
                y++;
            }
        }
        return cordinates;
    }

    public int getXLength() {
        int x = 0;
        for (Cordinate cordinate : cordinates) {
            if (cordinate.getX() > x) {
                x = cordinate.getX();
            }
        }
        return x;
    }

    public int getYLength() {
        int y = 0;
        for (Cordinate cordinate : cordinates) {
            if (cordinate.getY() > y) {
                y = cordinate.getY();
            }
        }
        return y;
    }
}

File: app/src/main/java/org/example/game/board/Board.kt
package org.example.game.board

import com.sun.scenario.effect.impl.sw.sse.SSEBlend_SRC_OUTPeer
import org.example.game.piece.Color
import org.example.game.piece.Piece
import java.lang.Math.abs

class Board (private var cordinates : List<Cordinate> ) {

    fun updateBoard(Cordinate1 : Cordinate, Cordinate2 : Cordinate) : Board{
        val mutableCordinates = mutableListOf<Cordinate>()
        for (i in cordinates){
            if (i.x == Cordinate1.x && i.y == Cordinate1.y){
                mutableCordinates.add(i.pieceDeparture())
                continue
            }
            if (i.x == Cordinate2.x && i.y == Cordinate2.y){
                mutableCordinates.add(i.pieceArrival(Cordinate1.piece!!))
                continue
            }
            mutableCordinates.add(i)
        }
        return Board(mutableCordinates)
    }

    fun getCordinatesInBetween(Cordinate1 : Cordinate, Cordinate2 : Cordinate) : List<Cordinate>{
        val cordinatesInBetween = mutableListOf<Cordinate>()
        val x1 = Cordinate1.x
        val y1 = Cordinate1.y
        val x2 = Cordinate2.x
        val y2 = Cordinate2.y
        if (x1 == x2) {
            val yMin = minOf(y1, y2)
            val yMax = maxOf(y1, y2)
            for (y in yMin + 1 until yMax) {
                cordinatesInBetween.add(getCordinate(x1, y))
            }
        } else if (y1 == y2) {
            val xMin = minOf(x1, x2)
            val xMax = maxOf(x1, x2)
            for (x in xMin + 1 until xMax) {
                cordinatesInBetween.add(getCordinate(x, y1))
            }
        }else{
            val xMin = minOf(x1, x2)
            val xMax = maxOf(x1, x2)
            val yMin = minOf(y1, y2)
            val yMax = maxOf(y1, y2)
            for (x in xMin + 1 until xMax){
                for (y in yMin + 1 until yMax){
                    if (Math.abs(x - x1) == Math.abs(y - y1)){
                        cordinatesInBetween.add(getCordinate(x , y))
                    }
                }
            }
        }
        return cordinatesInBetween
    }



    fun getCordinate(x: Int, y: Int): Cordinate {
        return cordinates.first { it.x == x && it.y == y }
    }

    fun getIdOfPiecesInBoard () : List<Int>{
        val idOfPieces = mutableListOf<Int>()
        cordinates.forEach {
            if (it.hasPiece()){
                idOfPieces.add(it.piece!!.id)
            }
        }
        return idOfPieces
    }

    fun getEnemyCordinates (color : Color) : List<Cordinate>{
        val enemyPieces = mutableListOf<Cordinate>()
        cordinates.forEach {
            if (it.hasPiece()){
                if (it.piece!!.color != color){
                    enemyPieces.add(it)
                }
            }
        }
        return enemyPieces
    }

    fun getAllyCordinates (color : Color) : List<Cordinate>{
        val allyPieces = mutableListOf<Cordinate>()
        cordinates.forEach {
            if (it.hasPiece()){
                if (it.piece!!.color == color){
                    allyPieces.add(it)
                }
            }
        }
        return allyPieces
    }

    fun getCordinateByPieceId (id : Int) : Cordinate {
        for (i in cordinates){
            if (i.hasPiece()){
                if (i.piece!!.id == id){
                    return i
                }
            }
        }
        throw Exception("No piece with id $id")
    }

    fun getCordinates() : List<Cordinate> {
        return cordinates
    }

    fun getBoardWidth() : Int {
        return cordinates.map { it.x }.max()!!
    }

    fun getBoardHeight() : Int {
        return cordinates.map { it.y }.max()!!
    }

    fun getAllPieces() : List<Piece> {
        val pieces = mutableListOf<Piece>()
        cordinates.forEach {
            if (it.hasPiece()){
                pieces.add(it.piece!!)
            }
        }
        return pieces
    }
}